README.md
# printer

golang版 终端打印机 

python版 终端打印机 [https://github.com/hellflame/terminal_printer](https://github.com/hellflame/terminal_printer)

### 效果展示

- 默认输出(第一次运行需要下载字体文件, 下载问题在后面说明)

![](./image/example1.png)

- 给定输出文字(字体库支持的文字)

![](./image/example2.png)

- 彩色图片输出

![](./image/example3.png)

- 图片灰度图(在终端分辨率较高的情况下输出)

![](./image/example4.png)

- 修改默认尺寸输出

![](./image/example5.png)

- 镂空效果 (交换填充字符)

![](./image/example6.png)

- 修改灰度边界阈值(`shade` 越大, 镂空区域判定越模糊, 图片输出有同样作用)

![](./image/example7.png)

![](./image/example8.png)

- 切换填充字符

![](./image/example9.png)

### 安装

```bash
go get -u github.com/hellflame/printer
```

若安装过程出现 `golang**/x/image` 下载失败等等连接失败的话有以下两种选择

##### 使用代理

可参考 https://goproxy.io/ (发现问题貌似依然存在)

##### 下载可执行的二进制文件(若只需要玩具本身)

查看 https://github.com/hellflame/printer/releases 

根据运行操作系统选择下载可执行文件，放在当前路径或可执行路径即可

### 打包发布

在支持 `make` 的环境中，只需 `make build` 或指定对应操作系统即可

```bash
make build
```

也可以手动打包支持当前操作系统

```bash
CGO_ENABLED=0 go build -ldflags "-s -w" 
```

则可在当前目录得到可执行文件

### 使用

```
terminal printer

Usage:
  printer [flags]

Flags:
  -c, --color int     color code, 30 ~ 50
  -f, --filter int    filter ascii code, 0 ~ 223 (default 73)
      --font string   font path or font index (default "0")
  -g, --gray          gray mode
  -e, --height uint   default console height (default 42)
  -h, --help          help for printer
  -r, --reverse       reverse back & foreground
  -s, --shade uint8   shade cliff (default 128)
  -t, --text string   render text content (default "hellflame")
      --version       version for printer
  -w, --width uint    default console width (default 158)
```

- `-t` 设置输出文字，默认作者名 (需要有字体，并且字体中包含文字)
- `-c` 设置颜色输出，默认 0 表示随机颜色
- `-f` 设置填充字符索引，部分索引输出为空白，默认输出 `i`
- `--font` 设置下载字体索引或给定字体路径 (图片不需要字体)
- `-g` 图片是否启用灰度模式 (文字输出默认灰度输出)
- `-s` 灰度滤波边界，即灰度二值化的边界值
- `-e` 设置输出高度，默认当前终端高度 (若无法获得，将使用一固定高度)
- `-w` 设置输出宽度，默认当前终端宽度 (若无法获得，将使用一固定宽度) 

部分使用示例在开头有展示

#### 代码复用

```go
import "github.com/hellflame/printer/draw"

func main(){
  img = draw.Clip(draw.Text(text, fontPath))
  result := draw.GeneratePixel(&img, filter, color, reverse?, gray?, shade)
  do_something_with(result)
}
```

### 字体下载

由于项目基于原python项目迁移，已存在字体库 https://github.com/hellflame/terminal_printer/tree/master/fonts

项目默认使用原项目字体路径，但只支持 `ttf` 字体

若自动下载字体缓慢，甚至下载失败，可选择手动下载字体文件，放在 `${HOME}/.terminal_fonts/` 即可

### License

The MIT License (MIT)

Copyright (c) 2019 hellflame

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


ChangeLog.md
## [0.1.3] - 2019-12-8

### Fix

- font index fix
- entry fix fatal output
- fix filter

### Add

- image preview

## [0.1.2] - 2019-12-7

### Add

- able to change gray shade cliff, default 128

### Change

- won't detect windows runtime directly, check stty output for reference

### Fix

- reverse mode bug fix
- some windows terminal support color
- build project with `-s -w` flags
- terminal color output suffix

## [0.1.0] - 2019-11-25

### Created

- migrate from python version of terminal printer https://github.com/hellflame/terminal_printer
- 取消了下载进度条监控


draw/init.go
package draw

import (
	"os"
	"os/exec"
	"os/user"
	"strconv"
	"strings"
)

const VERSION = "0.1.3"

var DefaultWidth = 50
var DefaultHeight = 80
var IsShabby = false
var FillBytes []int
var FontBase string
var DefaultFonts = []string{"shuyan.ttf", "letter.ttf",
	"Haibaoyuanyuan.ttf", "fengyun.ttf"}

const FontUrl = "https://raw.githubusercontent.com/hellflame/terminal_printer/" +
	"808004a7cd41b4383bfe6aa310c491c69d9b2556/fonts/"

const FillLength = 224

// decide default terminal size & initiate fill bytes & find out windows
func init() {
	cmd := exec.Command("stty", "size")
	cmd.Stdin = os.Stdin // this is important
	result, err := cmd.Output()
	if err == nil {
		parse := strings.Split(strings.TrimRight(string(result), "\n"),
			" ")
		DefaultHeight, _ = strconv.Atoi(parse[0])
		DefaultWidth, _ = strconv.Atoi(parse[1])
	} else {
		IsShabby = true
	}

	for i := 32; i < 32+FillLength; i++ {
		FillBytes = append(FillBytes, i)
	}

	usr, err := user.Current()
	var homeDIR string
	if err != nil {
		homeDIR = "."
	} else {
		homeDIR = usr.HomeDir
	}
	FontBase = homeDIR + "/.terminal_fonts/"
}


draw/text.go
// text drawer

package draw

import (
	"github.com/golang/freetype"
	"image"
	"image/draw"
	"io/ioutil"
	"log"
	"unicode/utf8"
)

// render text with given font
func Text(text string, fontPath string) *image.Gray {
	fontBytes, err := ioutil.ReadFile(fontPath)
	if err != nil {
		log.Fatal(err)
	}
	font, err := freetype.ParseFont(fontBytes)
	if err != nil {
		log.Fatal(err)
	}
	fontSize := 20.
	c := freetype.NewContext()
	c.SetDPI(96)
	c.SetFont(font)
	c.SetFontSize(fontSize)

	height := int(c.PointToFixed(fontSize) >> 6)
	width := utf8.RuneCountInString(text) * height // still need to crop

	fg, bg := image.Black, image.White

	canvas := image.NewGray(image.Rect(0, 0, width, height+5))
	draw.Draw(canvas, canvas.Bounds(), bg, image.Point{}, draw.Src)

	c.SetClip(canvas.Bounds())
	c.SetDst(canvas)
	c.SetSrc(fg)

	_, err = c.DrawString(text, freetype.Pt(0, height))
	if err != nil {
		log.Fatal(err)
	}

	return canvas
}

// clip text image to right size
// for golang doesn't have any direct way to get render string size
func Clip(img *image.Gray) image.Image {
	bound := img.Bounds().Max
	X, Y := bound.X, bound.Y

	var rowPixels [][]bool
	for y := 0; y < Y; y++ {
		var row []bool
		for x := 0; x < X; x++ {
			row = append(row, img.GrayAt(x, y).Y == 255)
		}
		rowPixels = append(rowPixels, row)
	}

	upBorder, upBoarderFound := 0, false
	downBorder := 0

	for y := 0; y < Y; y++ {
		isBlank := true
		for x := 0; x < X; x++ {
			if !rowPixels[y][x] {
				isBlank = false
			}
		}
		if !upBoarderFound && isBlank {
			upBorder++
		}
		if !isBlank {
			upBoarderFound = true
		}
		if downBorder == 0 && isBlank && upBoarderFound {
			downBorder = y
		}
	}

	leftBorder := 0
	rightBorder := X

	for x := 0; x < X; x++ {
		isBlank := true
		for y := 0; y < Y; y++ {
			if !rowPixels[y][x] {
				isBlank = false
			}
		}
		if !isBlank {
			break
		}

		if isBlank {
			leftBorder++
		}
	}
	for x := X - 1; x >= 0; x-- {
		isBlank := true
		for y := 0; y < Y; y++ {
			if !rowPixels[y][x] {
				isBlank = false
			}
		}
		if isBlank {
			rightBorder--
		}
		if !isBlank {
			break
		}

	}

	// possible issue
	if leftBorder > 0 {
		leftBorder -= 1
	}
	if downBorder <= 0 {
		downBorder = Y
	}
	rightBorder += 1

	return img.SubImage(image.Rect(leftBorder, upBorder, rightBorder, downBorder))
}


draw/console.go
// parse image to console

package draw

import (
	"fmt"
	"image"
	"math/rand"
	"strings"
	"time"
)

// generate terminal string
func GeneratePixel(img *image.Image, fillIndex int, colorCode int,
	reverse bool, grayMode bool, shadeCliff uint8) string {
	var renders []string
	src := *img
	bound := src.Bounds().Max
	X, Y := bound.X, bound.Y
	var fill uint8
	if fillIndex < 0 {
		fill = 4
	} else if fill > FillLength {
		fill = FillLength - 1
	} else {
		fill = uint8(fillIndex)
	}
	colorPool := rand.New(rand.NewSource(time.Now().UnixNano()))
	for y := 0; y < Y; y++ {
		var row []string
		for x := 0; x < X; x++ {
			var pix string
			r, g, b, _ := src.At(x, y).RGBA()
			if grayMode || IsShabby {
				shade := uint8((19595*r + 38470*g + 7471*b + 1<<15) >> 24) // convert to gray
				if reverse {
					if shade < shadeCliff {
						shade = 255 // image binary
					} else {
						shade = 0
					}
				} else {
					if shade < shadeCliff {
						shade = 0 // image binary
					} else {
						shade = 255
					}
				}
				if shade == 255 {
					pix = string(FillBytes[0])
				} else {
					pix = string(FillBytes[fill])
				}

			} else {
				pix = fmt.Sprintf("\033[0;38;2;%d;%d;%dm%s",
					r/257, g/257, b/257, string(FillBytes[fill]))
			}
			if colorCode == 0 && !IsShabby {
				// random color mode
				pix = fmt.Sprintf("\033[01;%dm%s", colorPool.Intn(10)+30, pix)
			}
			row = append(row, pix)
		}
		renders = append(renders, strings.Join(row, ""))
	}
	renderResult := strings.Join(renders, "\n")
	if colorCode > 0 {
		// specific color mode
		renderResult = fmt.Sprintf("\033[01;%dm", colorCode) + renderResult
	}
	// colorCode < 0
	return renderResult + "\033[00m"
}


entry.go
package main

import (
	"fmt"
	"github.com/hellflame/printer/draw"
	"github.com/nfnt/resize"
	"github.com/spf13/cobra"
	"image"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"strconv"
)

func exist(p string) bool {
	s, err := os.Stat(p)
	if err == nil {
		if !s.IsDir() {
			return true
		}
	}
	return false
}

func downloadFont(fontName string) {
	savePath := draw.FontBase + fontName
	if exist(savePath) {
		return
	}
	fmt.Printf("downloading font %s ...", fontName)
	resp, err := http.Get(draw.FontUrl + fontName)
	if err != nil {
		log.Fatal(err)
	}
	defer func() { _ = resp.Body.Close() }()
	stat, err := os.Stat(draw.FontBase)
	if err == nil {
		if !stat.IsDir() {
			log.Fatalf("save path is not dir, %s", draw.FontBase)
		}
	} else {
		if err = os.MkdirAll(draw.FontBase, 0755); err != nil {
			log.Fatal(err)
		}
	}
	content, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Fatal(err)
	}
	saver, err := os.Create(savePath)
	if err != nil {
		log.Fatal(err)
	}
	defer func() { _ = saver.Close() }()
	_, _ = saver.Write(content)
	fmt.Println("font saved")
}

func runner() {
	var imgPath string
	var run bool
	cmd := &cobra.Command{
		Use:     "printer",
		Short:   "terminal printer",
		Version: draw.VERSION,
		Run: func(cmd *cobra.Command, args []string) {
			if len(args) == 1 {
				imgPath = args[0]
			}
			run = true
		},
	}
	flag := cmd.Flags()
	text := flag.StringP("text", "t", "hellflame", "render text content")
	width := flag.UintP("width", "w", uint(draw.DefaultWidth), "default console width")
	height := flag.UintP("height", "e", uint(draw.DefaultHeight), "default console height")
	filter := flag.IntP("filter", "f", 73, fmt.Sprintf("filter ascii code, 0 ~ %d", draw.FillLength-1))
	color := flag.IntP("color", "c", 0, "color code, 30 ~ 50")
	gray := flag.BoolP("gray", "g", false, "gray mode")
	shade := flag.Uint8P("shade", "s", 128, "shade cliff")
	font := flag.String("font", "0", "font path or font index")
	reverse := flag.BoolP("reverse", "r", false, "reverse back & foreground")
	err := cmd.Execute()
	if err != nil {
		return
	}
	if !run {
		return
	}

	var img image.Image

	if imgPath != "" && exist(imgPath) {
		img = draw.LoadImage(imgPath)
	} else {
		fontIndex, err := strconv.Atoi(*font)
		if err != nil {
			// font is given by user
			if exist(*font) {
				img = draw.Clip(draw.Text(*text, *font))
			} else {
				fmt.Println("font path not exist")
				return
			}
		} else {
			if 0 <= fontIndex && fontIndex < len(draw.DefaultFonts) {
				// font is provide
				fontPath := draw.DefaultFonts[fontIndex]
				downloadFont(fontPath)
				img = draw.Clip(draw.Text(*text, draw.FontBase+fontPath))
			} else {
				fmt.Printf("font index should be 0 ~ %d\n", len(draw.DefaultFonts)-1)
				return
			}
		}
		*gray = true
	}

	img = resize.Resize(*width, *height, img, resize.Bilinear)
	fmt.Println(draw.GeneratePixel(&img, *filter, *color, *reverse, *gray, *shade))
}

func main() {
	runner()
}
